### PR TITLE
Palindromic Paths c++ solution

### PR DESCRIPTION
**Pull request by maggie-j-liu**

### CODE REVIEW

File: solutions/usaco-553.mdx [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+		prev[i][i] = 1;
+	}
+	for (int a = 1; a < n; a++)
+	{
+		/*
+		 * dp[i][j] is the number of palindromes with length 2a + 1
+		 * starting on row i and ending on row j
+		 * with the middle of the string on the diagonal
+		 */
+		long long dp[500][500]{};
```
Comment (dongliu0426) [inline solutions/usaco-553.mdx:48]: would prefer if you put this outside `for (a = 1...)` bc this reallocates `dp` every iteration

File: solutions/usaco-553.mdx [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+	freopen("palpath.in", "r", stdin);
+	freopen("palpath.out", "w", stdout);
+	int n;
+	cin >> n;
+	string grid[500];
+	for (int i = 0; i < n; i++)
+	{
+		cin >> grid[i];
+	}
+	long long prev[500][500]{};
```
Comment (SansPapyrus683) [inline solutions/usaco-553.mdx:36]: 1. plz use vectors
2. no magic numbers, maybe have a `MAX_N` or something

File: solutions/usaco-553.mdx [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+				if (col2 >= n)
+				{
+					continue;
+				}
+				/*
+				 * if the start and end of the string are not equal
+				 * it is not a palindrome so we continue
+				 */
+				if (grid[row1][col1] != grid[row2][col2])
+				{
```
Comment (SansPapyrus683) [inline solutions/usaco-553.mdx:68]: putting the opening brace on the same line would _really_ help with making it more concise

File: solutions/usaco-553.mdx [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+					dp[row1][row2] += prev[row1 + 1][row2 - 1];
+				}
+				dp[row1][row2] %= (int)(1e9 + 7);
+			}
+		}
+		// assign the current dp array to the prev array
+		swap(prev, dp);
+	}
+	cout << prev[0][n - 1] << endl;
+	return 0;
```
Comment (SansPapyrus683) [inline solutions/usaco-553.mdx:91]: return 0 isn't [required](https://stackoverflow.com/questions/4207134/what-is-the-proper-declaration-of-main-in-c)

File: solutions/usaco-553.mdx (hunk 1 above)
Comment (SansPapyrus683) [reply solutions/usaco-553.mdx:48]: i, think that's intended?

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]
idk man

File: solutions/usaco-553.mdx (hunk 1 above)
Comment (dongliu0426) [reply solutions/usaco-553.mdx:48]: use memset or smth

File: solutions/usaco-553.mdx (hunk 2 above)
Comment (maggie-j-liu) [reply solutions/usaco-553.mdx:36]: 1. why vectors?
2. sure

File: solutions/usaco-553.mdx (hunk 4 above)
Comment (maggie-j-liu) [reply solutions/usaco-553.mdx:91]: yeah, but is it wrong to have it?

Comment (SansPapyrus683) [reply solutions/usaco-553.mdx:91]: just 100% unnecessary

File: solutions/usaco-553.mdx (hunk 2 above)
Comment (SansPapyrus683) [reply solutions/usaco-553.mdx:36]: ~~arrays make me uncomfortable~~
if you're gonna use them make them global like everyone else does

File: solutions/usaco-553.mdx (hunk 1 above)
Comment (SansPapyrus683) [reply solutions/usaco-553.mdx:48]: that's a bit too advanced tbh

File: solutions/usaco-553.mdx (hunk 2 above)
Comment (maggie-j-liu) [reply solutions/usaco-553.mdx:36]: but there's nothing wrong with local arrays, right? is there a problem with memory?

File: solutions/usaco-553.mdx (hunk 1 above)
Comment (maggie-j-liu) [reply solutions/usaco-553.mdx:48]: i think reallocating dp every time is fine, it doesn't run slower or anything

unless there's something bad that could happen this way? idk

Comment (dongliu0426) [reply solutions/usaco-553.mdx]: alr

File: solutions/usaco-553.mdx (hunk 2 above)
Comment (SansPapyrus683) [reply solutions/usaco-553.mdx]: i,,,honestly don't know

Comment (maggie-j-liu) [reply solutions/usaco-553.mdx]: ok so this is fine

**Review by dongliu0426 — APPROVED** [review]

File: solutions/usaco-553.mdx [hunk 5]
```diff
@@ -0,0 +1,98 @@
+---
+id: usaco-553
+source: USACO Gold 2015 Open
+title: Palindromic Paths
+author: Maggie Liu
+---
+
+[Official Analysis](http://www.usaco.org/current/data/sol_palpath_gold.html)
+
+## Implementation
```
Comment (HarryWangATX) [inline solutions/usaco-553.mdx:10]: move inside cppsection and make it c++ imple

**Review by HarryWangATX — CHANGES_REQUESTED** [review]

File: solutions/usaco-553.mdx (hunk 5 above)
Comment (maggie-j-liu) [reply solutions/usaco-553.mdx:10]: this is the template on the guide, so i think this is fine

File: solutions/usaco-553.mdx [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+	}
+	for (int a = 1; a < n; a++)
+	{
+		/*
+		 * dp[i][j] is the number of palindromes with length 2a + 1
+		 * starting on row i and ending on row j
+		 * with the middle of the string on the diagonal
+		 */
+		long long dp[MAX_N][MAX_N]{};
+		for (int row1 = 0; row1 < n; row1++)
```
Comment (SansPapyrus683) [inline solutions/usaco-553.mdx:51]: maybe `start_row` idk what this `row1` is but you get the idea

File: solutions/usaco-553.mdx [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+	for (int i = 0; i < n; i++)
+	{
+		cin >> grid[i];
+	}
+	long long prev[MAX_N][MAX_N]{};
+	for (int i = 0; i < n; i++)
+	{
+		prev[i][i] = 1;
+	}
+	for (int a = 1; a < n; a++)
```
Comment (SansPapyrus683) [inline solutions/usaco-553.mdx:43]: `half_len` instead?

File: solutions/usaco-553.mdx [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+		 */
+		long long dp[MAX_N][MAX_N]{};
+		for (int row1 = 0; row1 < n; row1++)
+		{
+			int col1 = n - 1 - row1 - a;
+			if (col1 < 0)
+			{
+				continue;
+			}
+			for (int row2 = 0; row2 < n; row2++)
```
Comment (SansPapyrus683) [inline solutions/usaco-553.mdx:58]: and maybe make this `end_row`

File: solutions/usaco-553.mdx (hunk 7 above)
Comment (maggie-j-liu) [reply solutions/usaco-553.mdx:43]: the [official analysis](http://www.usaco.org/current/data/sol_palpath_gold.html) uses the variables a, r1, and r2 in their explanation, so i used a, row1, and row2 to be consistent while still being clear... if you really think i should change the name, i could do that, or i could add a comment

Comment (SansPapyrus683) [reply solutions/usaco-553.mdx:43]: ~~since when was the official analysis a good source of variable names~~

Comment (maggie-j-liu) [reply solutions/usaco-553.mdx:43]: just so it's not confusing when comparing with the official analysis... so it's easy to see at a glance which variables in the analysis correspond with which variables in the code

Comment (SansPapyrus683) [reply solutions/usaco-553.mdx:43]: then add comments plz

Comment (maggie-j-liu) [reply solutions/usaco-553.mdx:43]: I think the comment about the dp array describes the variables well (i forgot to change it when i renamed the variables but it should be good now) so idt i need to add another comment for each of the variables

Comment (andrewwangva) [reply solutions/usaco-553.mdx:43]: ^definitely not necessary to change variable names if the analysis uses it.

Comment (SansPapyrus683) [reply solutions/usaco-553.mdx:43]: ok yeah your new comment is p good

**Pull request merged**